File: frontend/src/components/benchmark/content/triangle_detection/implementations/triangle_detection.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class TriangleDetection
{
    private Dictionary<int, HashSet<int>> adjacencyList;

    public TriangleDetection()
    {
        adjacencyList = new Dictionary<int, HashSet<int>>();
    }

    public void AddEdge(int u, int v)
    {
        if (!adjacencyList.ContainsKey(u))
            adjacencyList[u] = new HashSet<int>();
        if (!adjacencyList.ContainsKey(v))
            adjacencyList[v] = new HashSet<int>();
        
        adjacencyList[u].Add(v);
        adjacencyList[v].Add(u);
    }

    public void BuildFromEdges(List<(int, int)> edges)
    {
        foreach (var (u, v) in edges)
        {
            AddEdge(u, v);
        }
    }

    public int CountTriangles()
    {
        int count = 0;
        foreach (var pair in adjacencyList)
        {
            int u = pair.Key;
            var neighbors = pair.Value;
            foreach (int v in neighbors)
            {
                if (v > u)
                {
                    foreach (int w in neighbors)
                    {
                        if (w > v && adjacencyList[v].Contains(w))
                        {
                            count++;
                        }
                    }
                }
            }
        }
        return count;
    }

    public List<(int, int, int)> FindTriangles()
    {
        var triangles = new List<(int, int, int)>();
        foreach (var pair in adjacencyList)
        {
            int u = pair.Key;
            var neighbors = pair.Value;
            foreach (int v in neighbors)
            {
                if (v > u)
                {
                    foreach (int w in neighbors)
                    {
                        if (w > v && adjacencyList[v].Contains(w))
                        {
                            triangles.Add((u, v, w));
                        }
                    }
                }
            }
        }
        return triangles;
    }

    public bool HasTriangle()
    {
        foreach (var pair in adjacencyList)
        {
            int u = pair.Key;
            var neighbors = pair.Value;
            foreach (int v in neighbors)
            {
                if (v > u)
                {
                    foreach (int w in neighbors)
                    {
                        if (w > v && adjacencyList[v].Contains(w))
                        {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }
}

File: frontend/src/components/benchmark/content/triangle_detection/implementations/triangle_detection.php
<?php

class TriangleDetection {
    private array $adjacencyList = [];

    public function __construct() {
        $this->adjacencyList = [];
    }

    public function addEdge(int $u, int $v): void {
        if (!isset($this->adjacencyList[$u])) {
            $this->adjacencyList[$u] = [];
        }
        if (!isset($this->adjacencyList[$v])) {
            $this->adjacencyList[$v] = [];
        }
        $this->adjacencyList[$u][$v] = true;
        $this->adjacencyList[$v][$u] = true;
    }

    public function buildFromEdges(array $edges): void {
        foreach ($edges as $edge) {
            $this->addEdge($edge[0], $edge[1]);
        }
    }

    public function countTriangles(): int {
        $count = 0;
        foreach ($this->adjacencyList as $u => $neighbors) {
            foreach (array_keys($neighbors) as $v) {
                if ($v > $u) {
                    foreach (array_keys($neighbors) as $w) {
                        if ($w > $v && isset($this->adjacencyList[$v][$w])) {
                            $count++;
                        }
                    }
                }
            }
        }
        return $count;
    }

    public function findTriangles(): array {
        $triangles = [];
        foreach ($this->adjacencyList as $u => $neighbors) {
            foreach (array_keys($neighbors) as $v) {
                if ($v > $u) {
                    foreach (array_keys($neighbors) as $w) {
                        if ($w > $v && isset($this->adjacencyList[$v][$w])) {
                            $triangles[] = [$u, $v, $w];
                        }
                    }
                }
            }
        }
        return $triangles;
    }

    public function hasTriangle(): bool {
        foreach ($this->adjacencyList as $u => $neighbors) {
            foreach (array_keys($neighbors) as $v) {
                if ($v > $u) {
                    foreach (array_keys($neighbors) as $w) {
                        if ($w > $v && isset($this->adjacencyList[$v][$w])) {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }
}

?>